src/Ganeti/HTools/Node.hs

```diff
@@ -1095,14 +1096,86 @@ availDisk t =
 iDsk :: Node -> Int
 iDsk t = truncate (tDsk t) - fDsk t
 
+-- | Returns state-of-world free memory on the node.
+-- | NOTE: This value is valid only before placement simulations.
+-- | TODO: Redefine this for memoy overcommitment.
+reportedFreeMem :: Node -> Int
+reportedFreeMem = fMem
+
+-- | Computes state-of-record free memory on the node.
+-- | TODO: Redefine this for memory overcommitment.
+recordedFreeMem :: Node -> Int
+recordedFreeMem t =
+  let total = tMem t
+      node = nMem t
+      inst = iMem t
+  in truncate total - node - inst
+
+-- | Computes the amount of missing memory on the node.
+-- NOTE: This formula uses free memory for calculations as opposed to
+-- used_memory in the definition, that's why it is the inverse.
+-- Explanations for missing memory (+) positive, (-) negative:
+-- (+) instances are using more memory that state-of-record
+--     - on KVM this might be due to the overhead per qemu process
+--     - on Xen manually upsized domains (xen mem-set)
+-- (+) on KVM non-qemu processes might be using more memory than what is
+--     reserved for node (no isolation)
+-- (-) on KVM qemu processes allocate memory on demand, thus an instance grows
+--     over its lifetime until it reaches state-of-record (+overhead)
+-- (-) on KVM KSM might be active
+-- (-) on Xen manually downsized domains (xen mem-set)
+missingMem :: Node -> Int
+missingMem t =
+  let state_of_world = reportedFreeMem t
+      state_of_record = recordedFreeMem t
+  in state_of_record - state_of_world
+
+-- | Computes the 'guaranteed' free memory, that is the minimum of what
+-- is reported by the node (available bytes) and our calculation based on
+-- instance sizes (our records), thus considering missing memory.
+-- NOTE 1: During placement simulations, the recorded memory changes, as
+-- instances are added/removed from the node, thus we have to calculate the
+-- missingMem (correction) before altering state-of-record and then
+-- use that correction to estimate state-of-world memory usage _after_
+-- the placements are done rather than doing min(record, world).
+-- NOTE 2: This is still only an approximation on KVM. As we shuffle instances
+-- during the simulation we are considering their state-of-record size, but
+-- in the real world the moves would shuffle parts of missing memory as well.
+-- Unfortunately as long as we don't have a more finegrained model that can
+-- better explain missing memory (split down based on root causes), we can't
+-- do better.
+-- NOTE 3: This is a hard limit based on available bytes and our bookkeeping.
+-- In case of memory overcommitment, both recordedFreeMem and reportedFreeMem
+-- would be extended by swap size on KVM or baloon size on Xen (their nominal
+-- and reported values).
+unallocatedMem :: Node -> Int
+unallocatedMem t =
```

Comment:
That's how I had it first. But that is wrong because of the same reasons xMem is part of the state not just dynamically calculated. As the state-of-record changes during instance shuffling simulations this would report incorrect values (see Note 1). And verifying this I see I had a bug there - instead of missingMem I should directly use xMem that was calculated at load time.